File: bluecross/input_inject.py
"""Input injection using uinput for virtual devices."""

import os
from typing import Optional

import evdev
from evdev import UInput, ecodes


class InputInjector:
    """Injects input events using uinput virtual devices."""
    
    def __init__(self):
        self._keyboard: Optional[UInput] = None
        self._mouse: Optional[UInput] = None
    
    def setup(self) -> None:
        """Set up virtual input devices."""
        # Create virtual keyboard with all standard keys
        key_caps = [
            code for code in range(ecodes.KEY_MAX)
            if code not in (ecodes.BTN_LEFT, ecodes.BTN_RIGHT, ecodes.BTN_MIDDLE,
                           ecodes.BTN_SIDE, ecodes.BTN_EXTRA)
        ]
        
        self._keyboard = UInput(
            {
                ecodes.EV_KEY: key_caps,
            },
            name="BlueCross Virtual Keyboard",
        )
        
        # Create virtual mouse
        self._mouse = UInput(
            {
                ecodes.EV_KEY: [
                    ecodes.BTN_LEFT,
                    ecodes.BTN_RIGHT,
                    ecodes.BTN_MIDDLE,
                    ecodes.BTN_SIDE,
                    ecodes.BTN_EXTRA,
                ],
                ecodes.EV_REL: [
                    ecodes.REL_X,
                    ecodes.REL_Y,
                    ecodes.REL_WHEEL,
                    ecodes.REL_HWHEEL,
                ],
            },
            name="BlueCross Virtual Mouse",
        )
        
        print("Virtual devices created:")
        print(f"  Keyboard: {self._keyboard.device.path}")
        print(f"  Mouse: {self._mouse.device.path}")
    
    def close(self) -> None:
        """Close virtual devices."""
        if self._keyboard:
            self._keyboard.close()
            self._keyboard = None
        if self._mouse:
            self._mouse.close()
            self._mouse = None
    
    def inject_key(self, code: int, value: int) -> None:
        """Inject a key event."""
        if not self._keyboard:
            return
        self._keyboard.write(ecodes.EV_KEY, code, value)
        self._keyboard.syn()
    
    def inject_mouse_move(self, dx: int, dy: int) -> None:
        """Inject relative mouse movement."""
        if not self._mouse:
            return
        if dx != 0:
            self._mouse.write(ecodes.EV_REL, ecodes.REL_X, dx)
        if dy != 0:
            self._mouse.write(ecodes.EV_REL, ecodes.REL_Y, dy)
        if dx != 0 or dy != 0:
            self._mouse.syn()
    
    def inject_mouse_button(self, button: int, value: int) -> None:
        """Inject a mouse button event."""
        if not self._mouse:
            return
        self._mouse.write(ecodes.EV_KEY, button, value)
        self._mouse.syn()
    
    def inject_mouse_scroll(self, dx: int, dy: int) -> None:
        """Inject mouse scroll events."""
        if not self._mouse:
            return
        if dy != 0:
            self._mouse.write(ecodes.EV_REL, ecodes.REL_WHEEL, dy)
        if dx != 0:
            self._mouse.write(ecodes.EV_REL, ecodes.REL_HWHEEL, dx)
        if dx != 0 or dy != 0:
            self._mouse.syn()

File: src/input_inject.rs
use evdev::uinput::VirtualDeviceBuilder;
use evdev::{AttributeSet, EventType, InputEvent, Key, RelativeAxisType};

pub struct InputInjector {
    keyboard: Option<evdev::uinput::VirtualDevice>,
    mouse: Option<evdev::uinput::VirtualDevice>,
}

impl InputInjector {
    pub fn new() -> Self {
        Self {
            keyboard: None,
            mouse: None,
        }
    }

    pub fn setup(&mut self) -> anyhow::Result<()> {
        // Create virtual keyboard with all standard keys (excluding mouse buttons)
        let mut keys = AttributeSet::<Key>::new();
        for code in 0..0x2ff_u16 {
            let key = Key::new(code);
            if key == Key::BTN_LEFT
                || key == Key::BTN_RIGHT
                || key == Key::BTN_MIDDLE
                || key == Key::BTN_SIDE
                || key == Key::BTN_EXTRA
            {
                continue;
            }
            keys.insert(key);
        }

        let keyboard = VirtualDeviceBuilder::new()?
            .name("BlueCross Virtual Keyboard")
            .with_keys(&keys)?
            .build()?;
        log::info!("Virtual keyboard created");
        self.keyboard = Some(keyboard);

        // Create virtual mouse
        let mut buttons = AttributeSet::<Key>::new();
        buttons.insert(Key::BTN_LEFT);
        buttons.insert(Key::BTN_RIGHT);
        buttons.insert(Key::BTN_MIDDLE);
        buttons.insert(Key::BTN_SIDE);
        buttons.insert(Key::BTN_EXTRA);

        let mut rel_axes = AttributeSet::<RelativeAxisType>::new();
        rel_axes.insert(RelativeAxisType::REL_X);
        rel_axes.insert(RelativeAxisType::REL_Y);
        rel_axes.insert(RelativeAxisType::REL_WHEEL);
        rel_axes.insert(RelativeAxisType::REL_HWHEEL);

        let mouse = VirtualDeviceBuilder::new()?
            .name("BlueCross Virtual Mouse")
            .with_keys(&buttons)?
            .with_relative_axes(&rel_axes)?
            .build()?;
        log::info!("Virtual mouse created");
        self.mouse = Some(mouse);

        Ok(())
    }

    pub fn inject_key(&mut self, code: u16, value: i32) {
        if let Some(ref mut kb) = self.keyboard {
            let events = [
                InputEvent::new(EventType::KEY, code, value),
                InputEvent::new(EventType::SYNCHRONIZATION, 0, 0),
            ];
            let _ = kb.emit(&events);
        }
    }

    pub fn inject_mouse_move(&mut self, dx: i32, dy: i32) {
        if let Some(ref mut mouse) = self.mouse {
            let mut events = Vec::with_capacity(3);
            if dx != 0 {
                events.push(InputEvent::new(
                    EventType::RELATIVE,
                    RelativeAxisType::REL_X.0,
                    dx,
                ));
            }
            if dy != 0 {
                events.push(InputEvent::new(
                    EventType::RELATIVE,
                    RelativeAxisType::REL_Y.0,
                    dy,
                ));
            }
            if !events.is_empty() {
                events.push(InputEvent::new(EventType::SYNCHRONIZATION, 0, 0));
                let _ = mouse.emit(&events);
            }
        }
    }

    pub fn inject_mouse_button(&mut self, button: u16, value: i32) {
        if let Some(ref mut mouse) = self.mouse {
            let events = [
                InputEvent::new(EventType::KEY, button, value),
                InputEvent::new(EventType::SYNCHRONIZATION, 0, 0),
            ];
            let _ = mouse.emit(&events);
        }
    }

    pub fn inject_mouse_scroll(&mut self, dx: i32, dy: i32) {
        if let Some(ref mut mouse) = self.mouse {
            let mut events = Vec::with_capacity(3);
            if dy != 0 {
                events.push(InputEvent::new(
                    EventType::RELATIVE,
                    RelativeAxisType::REL_WHEEL.0,
                    dy,
                ));
            }
            if dx != 0 {
                events.push(InputEvent::new(
                    EventType::RELATIVE,
                    RelativeAxisType::REL_HWHEEL.0,
                    dx,
                ));
            }
            if !events.is_empty() {
                events.push(InputEvent::new(EventType::SYNCHRONIZATION, 0, 0));
                let _ = mouse.emit(&events);
            }
        }
    }

    pub fn close(&mut self) {
        self.keyboard = None;
        self.mouse = None;
    }
}
